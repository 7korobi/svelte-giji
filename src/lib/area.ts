import type { SIZE, POINT, OFFSET } from './util'
import { isAndroid, isIOS, isLegacy, __BROWSER__, getVisualViewport, getInner } from './device'

type SIZE_WITH_SCALE = [number, number, number]

const vp = getVisualViewport()

class AreaBox {
  scale: number

  size: SIZE
  point!: POINT
  offset!: OFFSET

  keyboardHeight: number

  isZoom: boolean
  isKeyboard: boolean
  isPortrait!: boolean
  isLandscape!: boolean

  constructor([width, height, scale]: SIZE_WITH_SCALE, offset: OFFSET) {
    this.size = [width * scale, height * scale]

    this.scale = 1
    this.isZoom = false
    this.isKeyboard = false
    this.keyboardHeight = 0
    this.measureSize(width, height, 1)
    this.measureScroll(...offset)
  }

  measureSize(width: number, height: number, scale: number) {
    this.size = [width, height]
    this.scale = scale

    const isPortrait = width < height
    const isLandscape = height < width
    this.isPortrait = isPortrait
    this.isLandscape = isLandscape
  }
  measureScroll(top: number, right: number, bottom: number, left: number) {
    this.offset = [top, right, bottom, left]
    this.point = [left, top]
  }
}

const bootSize: SIZE_WITH_SCALE = [vp.width, vp.height, vp.scale]
const bootOffset: OFFSET = [0, 0, 0, 0]

let innerWidthOld = 0
let innerHeightOld = 0

export const ViewBox = new AreaBox(bootSize, bootOffset)
export const ZoomBox = new AreaBox(bootSize, bootOffset)
export const SafeAreaBox = new AreaBox(bootSize, bootOffset)

if (!isLegacy) {
  setResize()
  setScroll()
}

export function setScroll() {
  const { width, height, offsetLeft, offsetTop, pageLeft, pageTop } = getVisualViewport()
  const [vw, vh] = ViewBox.size

  const top = Math.ceil(offsetTop)
  const right = Math.floor(vw - width - offsetLeft)
  const bottom = Math.floor(vh - height - offsetTop)
  const left = Math.ceil(offsetLeft)

  ZoomBox.measureScroll(top, right, bottom, left)
  if (!ZoomBox.isZoom) {
    const pageRight = -pageLeft
    const pageBottom = -pageTop
    ViewBox.measureScroll(pageTop, pageRight, pageBottom, pageLeft)
  }
}

export function setResize() {
  const { width, height, scale } = getVisualViewport()
  ZoomBox.measureSize(width, height, scale)
  ViewFollowZoom()
  ViewFollowKbd(width, height)
  ViewFollowSize(width, height)
}

function ViewFollowZoom() {
  ViewBox.isZoom = ZoomBox.isZoom = chkZoom()
  // ズーム中であってもなくても、orientation change に追いつく処理だけはする。
  if (ViewBox.isPortrait !== ZoomBox.isPortrait && ViewBox.isLandscape !== ZoomBox.isLandscape) {
    const [height, width] = ViewBox.size
    ViewBox.size = [width, height]
    ViewBox.isPortrait = ZoomBox.isPortrait
    ViewBox.isLandscape = ZoomBox.isPortrait
  }
}

function chkZoom() {
  const { height, width, scale } = getVisualViewport()

  if (!isIOS) {
    return 1 < scale
  }

  // iOSのピンチは、操作中にズーム範囲を超えることがある。
  // iOSなので window.screen.availWidth はデバイスサイズ。
  // 1.0倍未満への縮小操作の検知にこの値との比較を使う。
  let { availHeight, availWidth } = window.screen
  if (
    (width < height && availWidth < availHeight) ||
    (width > height && availWidth > availHeight)
  ) {
  } else {
    availWidth = availHeight // swaped landscape.
  }
  return 1 < scale || availWidth < Math.floor(width)
}

function ViewFollowKbd(w2: number, h2: number) {
  const { innerWidth, innerHeight } = getInner()

  if (isIOS) {
    const kbdHeight = innerHeight - h2
    const isKbd = 80 < kbdHeight
    ViewBox.isKeyboard = ZoomBox.isKeyboard = isKbd
    ViewBox.keyboardHeight = ZoomBox.keyboardHeight = kbdHeight
  }
  if (isAndroid) {
    if (innerWidth === innerWidthOld) {
      if (innerHeightOld < innerHeight) {
        const kbdHeight = innerHeight - innerHeightOld
        if (80 < kbdHeight) {
          ViewBox.isKeyboard = ZoomBox.isKeyboard = false
          ViewBox.keyboardHeight = ZoomBox.keyboardHeight = kbdHeight
        }
      }
      if (innerHeight < innerHeightOld) {
        const kbdHeight = innerHeightOld - innerHeight
        if (80 < kbdHeight) {
          ViewBox.isKeyboard = ZoomBox.isKeyboard = true
          ViewBox.keyboardHeight = ZoomBox.keyboardHeight = kbdHeight
        }
      }
    }
  }
  innerWidthOld = innerWidth
  innerHeightOld = innerHeight
}

function ViewFollowSize(w2: number, h2: number) {
  const { isKeyboard, keyboardHeight } = ViewBox
  let [w1, h1] = ViewBox.size

  if (ZoomBox.isZoom) {
    h1 = (h2 * w1) / w2
  } else {
    w1 = w2
    h1 = h2
    if (isAndroid && isKeyboard) {
      h1 += keyboardHeight
    }
  }
  ViewBox.measureSize(w1, h1, 1)
}
