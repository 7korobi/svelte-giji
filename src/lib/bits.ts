const BITLIMIT = 31

type Label<T, U> = T | U | 'all'
type Labels<T, U> = readonly (T | U | 'all')[]
type BitsDic<T extends string, U extends string, X> = {
  [key in Label<T, U>]: X
}

export class BitsData<T extends string, U extends string | never> {
  value: number
  field: Bits<T, U>
  is: BitsDic<T, U, boolean>
  has: BitsDic<T, U, number>
  constructor(value: number, field: Bits<T, U>) {
    this.value = value
    this.field = field

    this.is = new Proxy(this, {
      get({ value, field }: BitsData<T, U>, label: Label<T, U>) {
        return Boolean(value & field.posi[label])
      },
      set(data: BitsData<T, U>, label: Label<T, U>, set: boolean) {
        const { value, field } = data
        const bits = set ? field.posi[label] : 0
        data.value = (value & field.nega[label]) | bits
        return true
      },
      has({ field }: BitsData<T, U>, label: Label<T, U>) {
        return !!field.idx[label]
      }
    }) as any

    this.has = new Proxy(this, {
      get({ value, field }: BitsData<T, U>, label: Label<T, U>) {
        return value & field.posi[label]
      },
      set(data: BitsData<T, U>, label: Label<T, U>, set: number) {
        const { value, field } = data
        data.value = (value & field.nega[label]) | (set & field.posi[label])
        return true
      },
      has({ field }: BitsData<T, U>, label: Label<T, U>) {
        return !!field.idx[label]
      }
    }) as any
  }

  posi(...labels: Labels<T, U>): BitsData<T, U> {
    let value = this.value
    const posi = this.field.posi
    for (const label of labels) {
      value |= posi[label]
    }
    return new BitsData<T, U>(value, this.field)
  }

  nega(...labels: Labels<T, U>): BitsData<T, U> {
    let value = this.value
    const nega = this.field.nega
    for (const label of labels) {
      value &= nega[label]
    }
    return new BitsData<T, U>(value, this.field)
  }

  toggle(...labels: Labels<T, U>): BitsData<T, U> {
    let value = this.value
    const posi = this.field.posi
    for (const label of labels) {
      value ^= posi[label]
    }
    return new BitsData<T, U>(value, this.field)
  }
}

export class Bits<T extends string, U extends string> {
  labels: Labels<T, U>
  mask: number
  posi: BitsDic<T, U, number>
  nega: BitsDic<T, U, number>
  idx: BitsDic<T, U, number>

  constructor(labels: readonly T[], options: { [key in U]: readonly T[] }) {
    if (BITLIMIT < labels.length) {
      throw new Error('too much bits.')
    }
    this.labels = labels
    this.mask = 2 ** labels.length - 1
    this.posi = {} as any
    this.nega = {} as any
    this.idx = {} as any
    labels.forEach(format.bind(this))
    labels.forEach(calc.bind(this))

    format.call(this, 'all')
    labels.forEach((key: T, idx) => {
      calc.call(this, 'all', idx)
    })

    for (const label in options) {
      format.call(this, label)
      options[label].forEach((key: T) => {
        calc.call(this, label, this.idx[key])
      })
      ;(this.labels as any[]).push(label)
    }
    ;(this.labels as any[]).push('all')

    function format(this: Bits<T, U>, label: Label<T, U>) {
      this.posi[label] = this.idx[label] = 0
      this.nega[label] = this.mask
    }
    function calc(this: Bits<T, U>, label: Label<T, U>, idx: number) {
      const posi = 2 ** idx
      const nega = this.mask & ~posi
      this.posi[label] |= posi
      this.nega[label] &= nega
      this.idx[label] || (this.idx[label] = idx)
    }
  }

  by(src: Labels<T, U>): number
  by(src: number): Labels<T, U>
  by(src: Labels<T, U> | number): Labels<T, U> | number {
    if ('number' === typeof src) {
      const labels: Labels<T, U> = []
      this.labels.forEach((label: Label<T, U>) => {
        const x = this.posi[label]
        if ((src & x) === x) {
          ;(labels as any[]).push(label)
        }
      })
      return labels
    }
    if (src instanceof Array) {
      let n = 0
      src.forEach((label: Label<T, U>) => {
        n |= (this.posi[label] || 0) as number
      })
      return n
    }
    throw new Error('invalid request type.')
  }

  data(n: number) {
    return new BitsData<T, U>(n, this)
  }

  to_str(n: number | BitsData<T, U>) {
    if (n instanceof BitsData) {
      n = n.value
    }
    return n.toString(36)
  }

  by_str(str: string | null | undefined): BitsData<T, U> {
    return this.data(str ? parseInt(str, 36) : 0)
  }

  to_url(n: number | BitsData<T, U>) {
    if (n instanceof BitsData) {
      n = n.value
    }
    return JSON.stringify(this.by(n))
  }

  static toggle(x: number, y: number) {
    if ((x & y) === y) {
      return x & ~y
    } else {
      return x | y
    }
  }

  static isSingle(x: number) {
    return 0 === (x & (x - 1))
  }

  static firstOff(x: number) {
    return x & (x - 1)
  }
  static firstOn(x: number) {
    return x | (x + 1)
  }

  static firstLinksOff(x: number) {
    return ((x | (x - 1)) + 1) & x
  }
  static firstLinksOn(x: number) {
    return ((x & (x + 1)) - 1) | x
  }

  static findBitOn(x: number) {
    return x & -x
  }
  static findBitOff(x: number) {
    return ~x & (x + 1)
  }

  static fillHeadsToOn(x: number) {
    return x | (x - 1)
  }

  static fillHeadsToOff(x: number) {
    return x & (x + 1)
  }

  static headsBitOff(x: number) {
    return ~x & (x - 1)
  }
  static headsBitOn(x: number) {
    return ~(~x | (x + 1))
  }

  static headsBitOffAndNextOn(x: number) {
    return x ^ (x - 1)
  }

  static snoob(x: number) {
    const minbit = x & -x
    const ripple = x + minbit
    const ones = ((x ^ ripple) >>> 2) / minbit
    return ripple | ones
  }

  static humming(x: number, y: number) {
    return this.count(x ^ y)
  }

  static count(x: number) {
    x = x - ((x >>> 1) & 0x55555555)
    x = (x & 0x33333333) + ((x >>> 2) & 0x33333333)
    x = (x + (x >>> 4)) & 0x0f0f0f0f
    x = x + (x >>> 8)
    x = x + (x >>> 16)
    return x & 0x3f // + x >> 32n
  }
}
