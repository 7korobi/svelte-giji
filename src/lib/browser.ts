import { readable, derived } from 'svelte/store'
import { __BROWSER__, addEventListener, removeEventListener } from './device'

function onLine() {
  if (__BROWSER__) {
    return window.navigator.onLine
  } else {
    return false
  }
}

function visibility() {
  if (__BROWSER__) {
    return document.visibilityState
  } else {
    return true
  }
}

export const isOnline = readable(onLine(), (set) => {
  addEventListener('offline', update)
  addEventListener('online', update)

  function update() {
    set(onLine())
    requestAnimationFrame(() => {
      set(onLine())
    })
  }

  return function release() {
    removeEventListener('offline', update)
    removeEventListener('online', update)
  }
})

export const isWatching = readable('hidden' !== visibility(), (set) => {
  addEventListener('visibilitychange', update)

  function update() {
    set('hidden' !== visibility())
    requestAnimationFrame(() => {
      set('hidden' !== visibility())
    })
  }

  return function release() {
    removeEventListener('visibilitychange', update)
  }
})

export const isActive = derived([isOnline, isWatching], ([online, watching]) => online && watching)

export function readyOnline(): Promise<void> {
  return new Promise((ok) => {
    if (onLine()) {
      ok()
    } else {
      addEventListener('online', update)
    }

    function update() {
      removeEventListener('online', update)
      ok()
    }
  })
}

export function readyWatching(): Promise<void> {
  return new Promise((ok) => {
    if ('hidden' !== visibility()) {
      ok()
    } else {
      addEventListener('visibilitychange', update)
    }

    function update() {
      if ('hidden' === visibility()) return
      removeEventListener('visibilitychange', update)
      ok()
    }
  })
}

export function readyDownload(
  el: HTMLImageElement | HTMLIFrameElement,
  url: string,
  timeout = 20000
): Promise<Event> {
  return new Promise((ok, ng) => {
    const timer = setTimeout(fail, timeout)
    el.addEventListener('--abort', fail)
    el.addEventListener('error', fail)
    el.addEventListener('load', success)
    el.src = url

    function bye() {
      clearTimeout(timer)
      el.removeEventListener('--abort', fail)
      el.removeEventListener('error', fail)
      el.removeEventListener('load', success)
    }

    function fail(e: Event = new addEventListener(`timeout ${timeout / 1000}sec`)) {
      el.src = ''
      bye()
      ng(e)
    }

    function success(e: Event) {
      bye()
      ok(e)
    }
  })
}
