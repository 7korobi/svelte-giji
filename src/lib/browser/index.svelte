<script lang="ts">
import { onDestroy } from 'svelte'

import Viewport from './viewport.svelte'
import Manifest from './manifest.svelte'
import SafeMeasure from './safe-measure.svelte'

import { __BROWSER__ } from './device'
import areaBoxInit from './area-box'
import browserInit from './browser'

onDestroy(areaBoxInit())
onDestroy(browserInit())

export let ratio: number
export let isDefaultSafeArea: boolean

// subscribe initial event.
</script>

<Manifest />
<Viewport min={1.0} max={1.0} />
<SafeMeasure bind:ratio bind:isDefaultSafeArea />
