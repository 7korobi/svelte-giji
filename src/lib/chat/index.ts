import Banner from './Banner.svelte'
import Talk from './Talk.svelte'
import Post from './Post.svelte'
import Report from './Report.svelte'

import Chat from './Chat.svelte'

export default Chat
export { Talk, Post, Report, Banner }