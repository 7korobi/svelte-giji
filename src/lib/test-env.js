window.indexedDB = require('fake-indexeddb')
window.IDBCursor = require('fake-indexeddb/lib/FDBCursor')
window.IDBCursorWithValue = require('fake-indexeddb/lib/FDBCursorWithValue')
window.IDBDatabase = require('fake-indexeddb/lib/FDBDatabase')
window.IDBFactory = require('fake-indexeddb/lib/FDBFactory')
window.IDBIndex = require('fake-indexeddb/lib/FDBIndex')
window.IDBKeyRange = require('fake-indexeddb/lib/FDBKeyRange')
window.IDBObjectStore = require('fake-indexeddb/lib/FDBObjectStore')
window.IDBOpenDBRequest = require('fake-indexeddb/lib/FDBOpenDBRequest')
window.IDBRequest = require('fake-indexeddb/lib/FDBRequest')
window.IDBTransaction = require('fake-indexeddb/lib/FDBTransaction')
window.IDBVersionChangeEvent = require('fake-indexeddb/lib/FDBVersionChangeEvent')
window.fetch = require('node-fetch')
