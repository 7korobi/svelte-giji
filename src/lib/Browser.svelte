<script lang="ts">
import { onMount } from 'svelte'
import { inner, outer, active } from './browser'
import { isRadius, isIOS, __BROWSER__ } from './device'

export let ratio = 1.0

const SAFE_WIDTH = 44
const SAFE_HEIGHT = 21

let isOnline = false
let isWatching = false
let innerWidth = 1000
let innerHeight = 1000
let outerWidth = 1000
let outerHeight = 1000
let safeAreaWidth = 1000
let safeAreaHeight = 1000
let safeAreaTop = 0
let safeAreaRight = 0
let safeAreaBottom = 0
let safeAreaLeft = 0

let measure: HTMLDivElement
$: css = measure && window.getComputedStyle(measure)

$: inner.update(() => [innerWidth, innerHeight, console.log("inner")])
$: outer.update(() => [outerWidth, outerHeight, console.log("outer")])

$: {
  const isActive = isOnline && isWatching
  active.update(() => ({ isActive, isOnline, isWatching }))
}

$: {
  if (css) {
    safeAreaTop = parseInt(css.marginTop)
    safeAreaRight = parseInt(css.marginRight)
    safeAreaBottom = parseInt(css.marginBottom)
    safeAreaLeft = parseInt(css.marginLeft)
    const zeroSafety =
      0.2 === Math.max(0.2, safeAreaTop, safeAreaRight, safeAreaBottom, safeAreaLeft)

    if (zeroSafety && isRadius && !isIOS) {
      if (outerHeight < outerWidth) {
        safeAreaLeft = safeAreaRight = SAFE_WIDTH
      }
      if (outerWidth < outerHeight) {
        safeAreaBottom = SAFE_HEIGHT
      }
    }
    safeAreaTop *= ratio
    safeAreaRight *= ratio
    safeAreaBottom *= ratio
    safeAreaLeft *= ratio
    safeAreaWidth = outerWidth - safeAreaLeft - safeAreaRight
    safeAreaHeight = outerHeight - safeAreaTop - safeAreaBottom
  }
}
onMount(() => {
  document.addEventListener('visibilitychange', updateWatching)
  updateWatching()

  return () => {
    document.removeEventListener('visibilitychange', updateWatching)
  }
})

function updateWatching() {
  isWatching = 'hidden' !== document.visibilityState
}
</script>

<svelte:window
  bind:innerWidth
  bind:innerHeight
  bind:outerWidth
  bind:outerHeight
  bind:online={isOnline} />

<div class="measure" bind:this={measure} />

<style>
.measure {
  pointer-events: none;
  user-select: none;
  visibility: hidden;

  z-index: -99999;

  opacity: 0;
  width: 0;
  height: 0;

  margin-top: env(safe-area-inset-top);
  margin-right: env(safe-area-inset-right);
  margin-bottom: env(safe-area-inset-bottom);
  margin-left: env(safe-area-inset-left);
}
</style>
