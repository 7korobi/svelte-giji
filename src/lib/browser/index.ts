import Browser from './main.svelte'
import Viewport from './viewport.svelte'
import KeyCapture from './key-capture.svelte'

import * as store from './store'

export * from './measure'
export { Browser, KeyCapture, Viewport }
export default store
