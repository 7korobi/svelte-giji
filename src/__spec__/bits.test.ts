import '../lib/test-env.js'
import { Bits } from '../lib/bits'

const TestBits = new Bits(['a', 'b', 'c', 'd', 'e', 'f', 'g'] as const, {
  a_c: ['a', 'b', 'c']
})
const ShowBits = new Bits(
  [
    'pin',
    'toc',
    'toc',
    'potof',
    'current',
    'search',
    'magnify',
    'side',
    'link',
    'mention'
  ] as const,
  {}
)

test('basic value', () => {
  const testValue = TestBits.by(TestBits.labels)
  const testBits = TestBits.data(testValue)
  expect(TestBits).toMatchSnapshot()
  expect(TestBits.labels).toEqual(TestBits.by(testValue))

  expect(testBits).toMatchSnapshot()
  expect(testBits.is.g).toEqual(true)
  expect(testBits.is.f).toEqual(true)
  expect(testBits.is.e).toEqual(true)
  expect(testBits.is.d).toEqual(true)
  expect(testBits.is.c).toEqual(true)
  expect(testBits.is.b).toEqual(true)
  expect(testBits.is.a).toEqual(true)
  expect(testBits.has.g).toEqual(64)
  expect(testBits.has.f).toEqual(32)
  expect(testBits.has.e).toEqual(16)
  expect(testBits.has.d).toEqual(8)
  expect(testBits.has.c).toEqual(4)
  expect(testBits.has.b).toEqual(2)
  expect(testBits.has.a).toEqual(1)
})

test('showlabel value', () => {
  const testValue = ShowBits.by(ShowBits.labels)
  const testBits = ShowBits.data(testValue)
  expect(ShowBits).toMatchSnapshot()
  expect(ShowBits.labels).toEqual(ShowBits.by(testValue))

  expect(testBits).toMatchSnapshot()
  expect(testBits.is.pin).toEqual(true)
  expect(testBits.is.toc).toEqual(true)
  expect(testBits.is.potof).toEqual(true)
  expect(testBits.is.current).toEqual(true)
  expect(testBits.is.search).toEqual(true)
  expect(testBits.is.magnify).toEqual(true)
  expect(testBits.is.side).toEqual(true)
  expect(testBits.is.link).toEqual(true)
  expect(testBits.is.mention).toEqual(true)
  expect((testBits.is as any).other).toEqual(false)
  expect(testBits.has.pin).toEqual(1)
  expect(testBits.has.toc).toEqual(6)
  expect(testBits.has.potof).toEqual(8)
  expect(testBits.has.current).toEqual(16)
  expect(testBits.has.search).toEqual(32)
  expect(testBits.has.magnify).toEqual(64)
  expect(testBits.has.side).toEqual(128)
  expect(testBits.has.link).toEqual(256)
  expect(testBits.has.mention).toEqual(512)
  expect((testBits.has as any).other).toEqual(0)
})

test('field calc', () => {
  expect(true).toEqual(Bits.isSingle(0b100000))
  expect(false).toEqual(Bits.isSingle(0b100100))
  expect(0b01100).toEqual(Bits.firstOff(0b01110))
  expect(0b01110).toEqual(Bits.firstOff(0b01111))
  expect(0b01101).toEqual(Bits.firstOn(0b01100))
  expect(0b00111).toEqual(Bits.firstOn(0b00011))
  expect(0b00010).toEqual(Bits.findBitOn(0b01110))
  expect(0b00010).toEqual(Bits.findBitOn(0b11110))
  expect(0b00001).toEqual(Bits.findBitOff(0b01100))
  expect(0b00100).toEqual(Bits.findBitOff(0b10011))
  expect(0b01111).toEqual(Bits.fillHeadsToOn(0b01100))
  expect(0b10011).toEqual(Bits.fillHeadsToOn(0b10011))
  expect(0b01100).toEqual(Bits.fillHeadsToOff(0b01100))
  expect(0b10000).toEqual(Bits.fillHeadsToOff(0b10011))
  expect(0b00000).toEqual(Bits.headsBitOn(0b01100))
  expect(0b00011).toEqual(Bits.headsBitOn(0b10011))
  expect(0b00011).toEqual(Bits.headsBitOff(0b01100))
  expect(0b00000).toEqual(Bits.headsBitOff(0b10011))
  expect(0b00111).toEqual(Bits.headsBitOffAndNextOn(0b01100))
  expect(0b00001).toEqual(Bits.headsBitOffAndNextOn(0b10011))

  expect(0b00010).toEqual(Bits.snoob(0b00001))
  expect(0b00110).toEqual(Bits.snoob(0b00101))

  expect(4).toEqual(Bits.count(0b110110))
  expect(3).toEqual(Bits.count(0b101100))
  expect(2).toEqual(Bits.count(0b000110))
})
