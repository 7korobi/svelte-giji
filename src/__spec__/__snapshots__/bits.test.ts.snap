// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic value 1`] = `
Bits {
  "idx": Object {
    "a": 0,
    "a_c": 1,
    "all": 1,
    "b": 1,
    "c": 2,
    "d": 3,
    "e": 4,
    "f": 5,
    "g": 6,
  },
  "labels": Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
    "a_c",
    "all",
  ],
  "mask": 127,
  "nega": Object {
    "a": 126,
    "a_c": 120,
    "all": 0,
    "b": 125,
    "c": 123,
    "d": 119,
    "e": 111,
    "f": 95,
    "g": 63,
  },
  "posi": Object {
    "a": 1,
    "a_c": 7,
    "all": 127,
    "b": 2,
    "c": 4,
    "d": 8,
    "e": 16,
    "f": 32,
    "g": 64,
  },
}
`;

exports[`basic value 2`] = `
BitsData {
  "field": Bits {
    "idx": Object {
      "a": 0,
      "a_c": 1,
      "all": 1,
      "b": 1,
      "c": 2,
      "d": 3,
      "e": 4,
      "f": 5,
      "g": 6,
    },
    "labels": Array [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
      "g",
      "a_c",
      "all",
    ],
    "mask": 127,
    "nega": Object {
      "a": 126,
      "a_c": 120,
      "all": 0,
      "b": 125,
      "c": 123,
      "d": 119,
      "e": 111,
      "f": 95,
      "g": 63,
    },
    "posi": Object {
      "a": 1,
      "a_c": 7,
      "all": 127,
      "b": 2,
      "c": 4,
      "d": 8,
      "e": 16,
      "f": 32,
      "g": 64,
    },
  },
  "has": Object {
    "field": 0,
    "has": 0,
    "is": 0,
    "value": 0,
  },
  "is": Object {
    "field": false,
    "has": false,
    "is": false,
    "value": false,
  },
  "value": 127,
}
`;

exports[`showlabel value 1`] = `
Bits {
  "idx": Object {
    "all": 1,
    "current": 4,
    "link": 8,
    "magnify": 6,
    "mention": 9,
    "pin": 0,
    "potof": 3,
    "search": 5,
    "side": 7,
    "toc": 1,
  },
  "labels": Array [
    "pin",
    "toc",
    "toc",
    "potof",
    "current",
    "search",
    "magnify",
    "side",
    "link",
    "mention",
    "all",
  ],
  "mask": 1023,
  "nega": Object {
    "all": 0,
    "current": 1007,
    "link": 767,
    "magnify": 959,
    "mention": 511,
    "pin": 1022,
    "potof": 1015,
    "search": 991,
    "side": 895,
    "toc": 1017,
  },
  "posi": Object {
    "all": 1023,
    "current": 16,
    "link": 256,
    "magnify": 64,
    "mention": 512,
    "pin": 1,
    "potof": 8,
    "search": 32,
    "side": 128,
    "toc": 6,
  },
}
`;

exports[`showlabel value 2`] = `
BitsData {
  "field": Bits {
    "idx": Object {
      "all": 1,
      "current": 4,
      "link": 8,
      "magnify": 6,
      "mention": 9,
      "pin": 0,
      "potof": 3,
      "search": 5,
      "side": 7,
      "toc": 1,
    },
    "labels": Array [
      "pin",
      "toc",
      "toc",
      "potof",
      "current",
      "search",
      "magnify",
      "side",
      "link",
      "mention",
      "all",
    ],
    "mask": 1023,
    "nega": Object {
      "all": 0,
      "current": 1007,
      "link": 767,
      "magnify": 959,
      "mention": 511,
      "pin": 1022,
      "potof": 1015,
      "search": 991,
      "side": 895,
      "toc": 1017,
    },
    "posi": Object {
      "all": 1023,
      "current": 16,
      "link": 256,
      "magnify": 64,
      "mention": 512,
      "pin": 1,
      "potof": 8,
      "search": 32,
      "side": 128,
      "toc": 6,
    },
  },
  "has": Object {
    "field": 0,
    "has": 0,
    "is": 0,
    "value": 0,
  },
  "is": Object {
    "field": false,
    "has": false,
    "is": false,
    "value": false,
  },
  "value": 1023,
}
`;
